File: packages/sdk/src/math/margin.ts
// Floating-point mirror of the on-chain margin math
// (programs/accelerated/src/utils/margin.rs). Use these for instant client-side quoting; the chain
// remains the source of truth.

import { BPS, type RiskTier, TIER_PARAMS } from '../constants.js';

export interface PositionQuote {
  /** Position notional in USDC. */
  notional: number;
  /** Position size in base token units. */
  size: number;
  /** Taker fee charged up front, in USDC. */
  fee: number;
  /** Realised leverage after fees. */
  effectiveLeverage: number;
  /** Price at which the position becomes liquidatable. */
  liquidationPrice: number;
}

const bps = (n: bigint) => Number(n) / Number(BPS);

/** Quote a new position the same way `open_position` sizes it on-chain. */
export function quotePosition(args: {
  collateral: number;
  leverage: number;
  price: number;
  isLong: boolean;
  tier: RiskTier;
  takerFeeBps?: number;
}): PositionQuote {
  const { collateral, leverage, price, isLong, tier } = args;
  const takerFeeBps = args.takerFeeBps ?? 25;
  const mmr = bps(TIER_PARAMS[tier].maintenanceMarginBps);

  const fee = (collateral * leverage * takerFeeBps) / 10_000;
  const collateralAfterFee = Math.max(0, collateral - fee);
  const notional = collateralAfterFee * leverage;
  const size = price > 0 ? notional / price : 0;
  const effectiveLeverage = collateralAfterFee > 0 ? notional / collateralAfterFee : 0;

  const maintenance = notional * mmr;
  const buffer = Math.max(0, collateralAfterFee - maintenance);
  const priceDelta = size > 0 ? buffer / size : 0;
  const liquidationPrice = isLong ? Math.max(0, price - priceDelta) : price + priceDelta;

  return { notional, size, fee, effectiveLeverage, liquidationPrice };
}

/** Margin ratio in bps, matching `margin_ratio_bps` on-chain. */
export function marginRatioBps(args: {
  collateral: number;
  size: number;
  entryPrice: number;
  markPrice: number;
  isLong: boolean;
  fundingOwed?: number;
}): number {
  const { collateral, size, entryPrice, markPrice, isLong } = args;
  const fundingOwed = args.fundingOwed ?? 0;
  const pnl = isLong ? size * (markPrice - entryPrice) : size * (entryPrice - markPrice);
  const equity = Math.max(0, collateral + pnl - fundingOwed);
  const notional = size * markPrice;
  if (notional === 0) return Number(BPS);
  return (equity / notional) * Number(BPS);
}

/** Whether a position can be liquidated at the given mark. */
export function isLiquidatable(marginRatioBps: number, tier: RiskTier): boolean {
  return marginRatioBps <= Number(TIER_PARAMS[tier].maintenanceMarginBps);
}

File: programs/accelerated/src/utils/margin.rs
//! Margin and liquidation math.
//!
//! A position's *equity* is its posted collateral plus unrealized PnL minus accrued funding.
//! The *margin ratio* is equity / notional. When the margin ratio falls below the market's
//! maintenance-margin requirement, the position is liquidatable by any keeper.

use crate::constants::{BPS_DENOMINATOR, PRICE_PRECISION};
use crate::errors::AcceleratedError;
use crate::utils::math::{notional, unrealized_pnl, SafeMath};
use anchor_lang::prelude::*;

/// Account equity in USDC base units (saturating at zero — a position can be underwater but never
/// negative equity from the trader's perspective; the deficit is the vault's loss).
pub fn equity(collateral: u64, pnl: i128, funding_owed: i128) -> Result<u64> {
    let base = collateral as i128;
    let eq = base.safe_add(pnl)?.safe_sub(funding_owed)?;
    Ok(if eq < 0 { 0 } else { eq as u64 })
}

/// Margin ratio in basis points: equity / notional * 10_000.
pub fn margin_ratio_bps(
    collateral: u64,
    base_size: u128,
    entry_price: u128,
    mark_price: u128,
    is_long: bool,
    funding_owed: i128,
) -> Result<u64> {
    let pnl = unrealized_pnl(base_size, entry_price, mark_price, is_long)?;
    let eq = equity(collateral, pnl, funding_owed)? as u128;
    let notional_now = notional(base_size, mark_price)?;
    if notional_now == 0 {
        return Ok(BPS_DENOMINATOR);
    }
    let ratio = eq.safe_mul(BPS_DENOMINATOR as u128)?.safe_div(notional_now)?;
    Ok(u64::try_from(ratio).unwrap_or(u64::MAX))
}

/// True when the position's margin ratio is at or below the maintenance requirement.
pub fn is_liquidatable(margin_ratio_bps: u64, maintenance_margin_bps: u64) -> bool {
    margin_ratio_bps <= maintenance_margin_bps
}

/// Effective leverage in bps at entry: notional / collateral * 10_000.
pub fn entry_leverage_bps(notional_usdc: u128, collateral: u64) -> Result<u64> {
    if collateral == 0 {
        return err!(AcceleratedError::CollateralTooLow);
    }
    let lev = notional_usdc.safe_mul(BPS_DENOMINATOR as u128)?.safe_div(collateral as u128)?;
    Ok(u64::try_from(lev).unwrap_or(u64::MAX))
}

/// The price at which a position's equity hits the maintenance threshold.
///
/// For a long:  liq = entry * (1 - 1/leverage + mmr)
/// For a short: liq = entry * (1 + 1/leverage - mmr)
/// Returned in PRICE_PRECISION units; 0 means "effectively unliquidatable in range".
pub fn liquidation_price(
    entry_price: u128,
    collateral: u64,
    base_size: u128,
    maintenance_margin_bps: u64,
    is_long: bool,
) -> Result<u128> {
    let notional_at_entry = notional(base_size, entry_price)?;
    if notional_at_entry == 0 {
        return Ok(0);
    }
    // maintenance requirement in USDC
    let maint = notional_at_entry
        .safe_mul(maintenance_margin_bps as u128)?
        .safe_div(BPS_DENOMINATOR as u128)?;
    // price move (in price units) that exhausts collateral down to maintenance
    let buffer = (collateral as u128).safe_sub(maint).unwrap_or(0);
    let price_delta = buffer.safe_mul(PRICE_PRECISION)?.safe_div(base_size)?;
    if is_long {
        Ok(entry_price.safe_sub(price_delta).unwrap_or(0))
    } else {
        entry_price.safe_add(price_delta)
    }
}
